keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,802 @@
+
+In theory, `KUBE-MARK-DROP` is useful for the same reason as
+`KUBE-MARK-MASQ`, although in practice, it turns out to be
+unnecessary. Kube-proxy currently drops packets in two cases:
+
+  - When using the `LoadBalancerSourceRanges` feature on a service,
+    any packet to that service that comes from outside the valid
+    source IP ranges is dropped.
+
+  - When using `Local` external traffic policy, if a connection to a
+    service arrives on a node that has no local endpoints for that
+    service, it is dropped.
+
+In the `LoadBalancerSourceRanges` case, it would not be difficult to
+avoid using `KUBE-MARK-DROP`. The current rule logic creates a
+per-service `KUBE-FW-` chain that checks against each allowed source
+range and forwards traffic from those ranges to the `KUBE-SVC-` or
+`KUBE-XLB-` chain for the service. At the end of the chain, it calls
+`KUBE-MARK-DROP` on any unmatched packets.
+
+To do this without `KUBE-MARK-DROP`, we simply remove the
+`KUBE-MARK-DROP` call at the end of the `KUBE-FW-` chain, and add a
+new rule in the `filter` table, matching on the load balancer IP and
+port, and calling `-j DROP`. Any traffic that matched one of the
+allowed source ranges in the `KUBE-FW-` chain would have been DNAT'ed
+to a pod IP by this point, so any packet that still has the original
+load balancer IP as its destination when it reaches the `filter` table
+must be a packet that was not accepted, so we can drop it.
+
+In the external traffic policy case, things are even simpler; we can
+just move the existing traffic-dropping rule from the `nat` table to
+the `filter` table and call `DROP` directly rather than
+`KUBE-MARK-DROP`, and we will get exactly the same effect. (Indeed, we
+already do it that way when calling `REJECT`, so this will actually
+make things more consistent.)
+
+#### External Users of Kubelet's IPTables Chains
+
+Some users or third-party software may expect that certain IPTables
+chains always exist on nodes in a Kubernetes cluster.
+
+For example, the [CNI portmap plugin] provides an
+`"externalSetMarkChain"` option that is explicitly intended to be used
+with `"KUBE-MARK-MASQ"`, to make the plugin use Kubernetes's iptables
+rules instead of creating its own. Although in most cases kube-proxy
+will also create `KUBE-MARK-MASQ`, kube-proxy may not start up until
+after some other components, and some users may be running a network
+plugin that has its own proxy implementation rather than using
+kube-proxy. Thus, users may end up in a situation where some software
+is trying to use a `KUBE-MARK-MASQ` chain that does not exist.
+
+Most of these external components have simply copied kube-proxy's use
+of `KUBE-MARK-MASQ` without understanding why it is used that way in
+kube-proxy. But because they are generally doing much less IPTables
+processing than kube-proxy does, they could fairly easily be rewritten
+to not use the packet mark, and just have slightly-redundant
+`PREROUTING` and `POSTROUTING` rules instead.
```

Comment:
Yeah, we can fix CNI to not use a mark, though it will have a lot of duplicated logic.